packages/firestore/src/local/indexeddb_remote_document_cache.ts

```diff
@@ -331,95 +295,62 @@ export class IndexedDbRemoteDocumentCache implements RemoteDocumentCache {
-          documentPromises.push(
-            this.getEntry(transaction, key).next(maybeDocument => {
-              const doc =
-                maybeDocument ||
-                new NoDocument(key, SnapshotVersion.forDeletedDoc());
-              changedDocs = changedDocs.insert(key, doc);
-            })
+    const documentsStore = remoteDocumentsStore(transaction);
+    const range = IDBKeyRange.lowerBound(lastReadTime, true);
+    return documentsStore
+      .iterate(
+        { index: DbRemoteDocument.readTimeIndex, range },
+        (_, dbRemoteDoc) => {
+          const doc = this.serializer.fromDbRemoteDocument(dbRemoteDoc);
+          changedDocs = changedDocs.insert(doc.key, doc);
+          this.lastProcessedReadTime = this.serializer.fromDbTimestampKey(
+            dbRemoteDoc.readTime!
           );
-        });
-        return PersistencePromise.waitFor(documentPromises);
-      })
+        }
+      )
       .next(() => changedDocs);
   }
 
   /**
-   * Removes all changes in the remote document changelog through `changeId`
-   * (inclusive).
+   * Sets the last processed read time to the maximum read time from cache,
+   * allowing calls to getNewDocumentChanges() to return subsequent changes.
    */
-  removeDocumentChangesThroughChangeId(
-    transaction: PersistenceTransaction,
-    changeId: number
-  ): PersistencePromise<void> {
-    const range = IDBKeyRange.upperBound(changeId);
-    return documentChangesStore(transaction).delete(range);
-  }
-
-  private synchronizeLastDocumentChangeId(
-    documentChangesStore: SimpleDbStore<
-      DbRemoteDocumentChangesKey,
-      DbRemoteDocumentChanges
-    >
+  private synchronizeLastProcessedReadTime(
+    transaction: SimpleDbTransaction
   ): PersistencePromise<void> {
-    // If there are no existing changes, we set `lastProcessedDocumentChangeId`
-    // to 0 since IndexedDb's auto-generated keys start at 1.
-    this._lastProcessedDocumentChangeId = 0;
-    return documentChangesStore.iterate(
-      { keysOnly: true, reverse: true },
+    const documentsStore = SimpleDb.getStore<
+      DbRemoteDocumentKey,
+      DbRemoteDocument
+    >(transaction, DbRemoteDocument.store);
+
+    // If there are no existing entries, we set `lastProcessedReadTime` to 0.
+    this.lastProcessedReadTime = SnapshotVersion.forDeletedDoc();
+    return documentsStore.iterate(
+      { index: DbRemoteDocument.readTimeIndex, reverse: true },
       (key, value, control) => {
-        this._lastProcessedDocumentChangeId = key;
+        if (value.readTime) {
+          this.lastProcessedReadTime = SnapshotVersion.fromTimestamp(
+            new Timestamp(value.readTime![0], value.readTime![1])
+          );
+        }
         control.done();
       }
     );
   }
 
-  newChangeBuffer(): RemoteDocumentChangeBuffer {
-    return new IndexedDbRemoteDocumentCache.RemoteDocumentChangeBuffer(this);
+  newChangeBuffer(options: {
+    createSentinelDocumentsToTrackDeletes: boolean;
```

Comment:
1. You say "sentinelDocuments" here but refer to them as "sentinel deletes" elsewhere.  Please use consistent terminology.
2. We have this long name here but just `trackDeletes` in the RemoteDocumentChangeBuffer constructor.  I am not sure which I prefer, but I can't think of a great reason for them to be different.